File: analyzer.py
# analyzer.py - symmetry analysis using grid transforms
import numpy as np

def reflect_grid(grid, axis='horizontal'):
    if axis == 'horizontal':
        return grid[::-1]
    elif axis == 'vertical':
        return [row[::-1] for row in grid]
    elif axis == 'transpose':
        # convert list of rows of points to transposed structure (may be irregular)
        # we approximate transpose by using column indices up to min row length
        nrows = len(grid)
        ncols = min(len(r) for r in grid)
        out = []
        for c in range(ncols):
            col = []
            for r in range(nrows):
                col.append(grid[r][c])
            out.append(col)
        return out
    return grid

def rotate_grid_90(grid):
    # rotate points by 90 degrees around centroid (approx) and re-cluster is complex;
    # here we rotate the grid indices (not pixel coords) if grid is regular.
    transposed = reflect_grid(grid, 'transpose')
    return [list(reversed(r)) for r in transposed]

def grids_equal(g1, g2, tol=12):
    if len(g1) != len(g2):
        return False
    for row1, row2 in zip(g1, g2):
        if len(row1) != len(row2):
            return False
        for p1, p2 in zip(row1, row2):
            if abs(p1[0] - p2[0]) > tol or abs(p1[1] - p2[1]) > tol:
                return False
    return True

def analyze_symmetries(dot_grid):
    results = {}
    results['horizontal_reflection'] = grids_equal(dot_grid, reflect_grid(dot_grid, 'horizontal'))
    results['vertical_reflection'] = grids_equal(dot_grid, reflect_grid(dot_grid, 'vertical'))
    try:
        results['rotation_90'] = grids_equal(dot_grid, rotate_grid_90(dot_grid))
        results['rotation_180'] = grids_equal(dot_grid, rotate_grid_90(rotate_grid_90(dot_grid)))
    except Exception:
        results['rotation_90'] = False
        results['rotation_180'] = False
    return results

File: analyzer.js
// analyzer.js - Symmetry analysis for kolam patterns

class Analyzer {
    static reflectGrid(grid, axis = 'horizontal') {
        if (!grid || grid.length === 0) return grid;

        switch (axis) {
            case 'horizontal':
                return [...grid].reverse();
            case 'vertical':
                return grid.map(row => [...row].reverse());
            case 'transpose':
                const maxCols = Math.max(...grid.map(row => row.length));
                const transposed = [];
                for (let col = 0; col < maxCols; col++) {
                    const newRow = [];
                    for (let row = 0; row < grid.length; row++) {
                        if (col < grid[row].length) {
                            newRow.push(grid[row][col]);
                        }
                    }
                    if (newRow.length > 0) {
                        transposed.push(newRow);
                    }
                }
                return transposed;
            default:
                return grid;
        }
    }

    static rotateGrid90(grid) {
        if (!grid || grid.length === 0) return grid;
        
        const transposed = this.reflectGrid(grid, 'transpose');
        return transposed.map(row => [...row].reverse());
    }

    static gridsEqual(g1, g2, tolerance = 15) {
        if (!g1 || !g2 || g1.length !== g2.length) return false;

        for (let i = 0; i < g1.length; i++) {
            if (g1[i].length !== g2[i].length) return false;
            
            for (let j = 0; j < g1[i].length; j++) {
                const [x1, y1] = g1[i][j];
                const [x2, y2] = g2[i][j];
                
                if (Math.abs(x1 - x2) > tolerance || Math.abs(y1 - y2) > tolerance) {
                    return false;
                }
            }
        }
        return true;
    }

    static analyzeSymmetries(grid) {
        if (!grid || grid.length === 0) {
            return {
                horizontal_reflection: false,
                vertical_reflection: false,
                rotation_90: false,
                rotation_180: false
            };
        }

        const results = {};
        
        try {
            // Horizontal reflection (flip top-bottom)
            const horizontalReflected = this.reflectGrid(grid, 'horizontal');
            results.horizontal_reflection = this.gridsEqual(grid, horizontalReflected);

            // Vertical reflection (flip left-right)
            const verticalReflected = this.reflectGrid(grid, 'vertical');
            results.vertical_reflection = this.gridsEqual(grid, verticalReflected);

            // 90-degree rotation
            const rotated90 = this.rotateGrid90(grid);
            results.rotation_90 = this.gridsEqual(grid, rotated90);

            // 180-degree rotation
            const rotated180 = this.rotateGrid90(this.rotateGrid90(grid));
            results.rotation_180 = this.gridsEqual(grid, rotated180);

        } catch (error) {
            console.warn('Error in symmetry analysis:', error);
            results.horizontal_reflection = false;
            results.vertical_reflection = false;
            results.rotation_90 = false;
            results.rotation_180 = false;
        }

        return results;
    }

    static getSymmetryScore(symmetries) {
        const trueCount = Object.values(symmetries).filter(Boolean).length;
        return trueCount / Object.keys(symmetries).length;
    }

    static classifyPattern(grid, symmetries) {
        const score = this.getSymmetryScore(symmetries);
        const totalDots = grid.reduce((sum, row) => sum + row.length, 0);

        if (score >= 0.75) {
            return 'symmetric';
        } else if (score >= 0.25) {
            return 'tiling';
        } else {
            return 'freeform';
        }
    }
}